File: smash/solver/module/mwd_parameters.f90
!%      This module `mwd_parameters` encapsulates all SMASH parameters.
!%      This module is wrapped and differentiated.
!%
!%      ParametersDT type:
!%      
!%      </> Public
!%      ======================== =======================================
!%      `Variables`              Description
!%      ======================== =======================================
!%      ``ci``                   Interception parameter          [mm]    (default: 1)     ]0, +Inf[
!%      ``cp``                   Production parameter            [mm]    (default: 200)   ]0, +Inf[
!%      ``beta``                 Percolation parameter           [-]     (default: 1000)  ]0, +Inf[
!%      ``cft``                  Fast transfer parameter         [mm]    (default: 500)   ]0, +Inf[
!%      ``cst``                  Slow transfer parameter         [mm]    (default: 500)   ]0, +Inf[
!%      ``alpha``                Transfer partitioning parameter [-]     (default: 0.9)   ]0, 1[
!%      ``exc``                  Exchange parameter              [mm/dt] (default: 0)     ]-Inf, +Inf[
!%      ``lr``                   Linear routing parameter        [min]   (default: 5)     ]0, +Inf[
!%      ======================== =======================================
!%
!%      contains
!%
!%      [1] ParametersDT_initialise
!%      [2] parameters_to_matrix
!%      [3] matrix_to_parameters
!%      [4] vector_to_parameters
!%      [5] set0_parameters
!%      [6] set1_parameters

module mwd_parameters

    use mwd_common !% only: sp, np
    use mwd_mesh  !% only: MeshDT
    use mwd_input_data !% only: Input_DataDT
    
    implicit none
    
    type ParametersDT
        
        real(sp), dimension(:,:), allocatable :: ci
        real(sp), dimension(:,:), allocatable :: cp
        real(sp), dimension(:,:), allocatable :: beta
        real(sp), dimension(:,:), allocatable :: cft
        real(sp), dimension(:,:), allocatable :: cst
        real(sp), dimension(:,:), allocatable :: alpha
        real(sp), dimension(:,:), allocatable :: exc
        real(sp), dimension(:,:), allocatable :: lr
        
    end type ParametersDT
    
    type Hyper_ParametersDT
    
        real(sp), dimension(:,:), allocatable :: ci
        real(sp), dimension(:,:), allocatable :: cp
        real(sp), dimension(:,:), allocatable :: beta
        real(sp), dimension(:,:), allocatable :: cft
        real(sp), dimension(:,:), allocatable :: cst
        real(sp), dimension(:,:), allocatable :: alpha
        real(sp), dimension(:,:), allocatable :: exc
        real(sp), dimension(:,:), allocatable :: lr
    
    end type Hyper_ParametersDT
    
    contains
        
        subroutine ParametersDT_initialise(parameters, mesh)
        
            !% Notes
            !% -----
            !%
            !% ParametersDT initialisation subroutine
        
            implicit none
            
            type(MeshDT), intent(in) :: mesh
            type(ParametersDT), intent(inout) :: parameters
            
            integer :: nrow, ncol
            
            nrow = mesh%nrow
            ncol = mesh%ncol
            
            allocate(parameters%ci(nrow, ncol))
            allocate(parameters%cp(nrow, ncol))
            allocate(parameters%beta(nrow, ncol))
            allocate(parameters%cft(nrow, ncol))
            allocate(parameters%cst(nrow, ncol))
            allocate(parameters%alpha(nrow, ncol))
            allocate(parameters%exc(nrow, ncol))
            allocate(parameters%lr(nrow, ncol))
            
            parameters%ci    = 1._sp
            parameters%cp    = 200._sp
            parameters%beta  = 1000._sp
            parameters%cft   = 500._sp
            parameters%cst   = 500._sp
            parameters%alpha = 0.9_sp
            parameters%exc   = 0._sp
            parameters%lr    = 5._sp
 
        end subroutine ParametersDT_initialise
        
        
        subroutine Hyper_ParametersDT_initialise(hyper_parameters, setup, mesh)
        
            !% Notes
            !% -----
            !%
            !% ParametersDT initialisation subroutine
        
            implicit none
            
            type(SetupDT), intent(in) :: setup
            type(MeshDT), intent(in) :: mesh
            type(Hyper_ParametersDT), intent(inout) :: hyper_parameters
            
            integer :: n
            
            n = np * (1 + 2 * setup%nd)
            
            allocate(hyper_parameters%ci(n, 1))
            allocate(hyper_parameters%cp(n, 1))
            allocate(hyper_parameters%beta(n, 1))
            allocate(hyper_parameters%cft(n, 1))
            allocate(hyper_parameters%cst(n, 1))
            allocate(hyper_parameters%alpha(n, 1))
            allocate(hyper_parameters%exc(n, 1))
            allocate(hyper_parameters%lr(n, 1))
            
            hyper_parameters%ci    = 0._sp
            hyper_parameters%cp    = 0._sp
            hyper_parameters%beta  = 0._sp
            hyper_parameters%cft   = 0._sp
            hyper_parameters%cst   = 0._sp
            hyper_parameters%alpha = 0._sp
            hyper_parameters%exc   = 0._sp
            hyper_parameters%lr    = 0._sp
 
        end subroutine Hyper_ParametersDT_initialise
        
!%      TODO comment  
        subroutine parameters_to_matrix(parameters, matrix)
        
            implicit none
            
            type(ParametersDT), intent(in) :: parameters
            real(sp), dimension(size(parameters%cp, 1), &
            & size(parameters%cp, 2), np), intent(inout) :: matrix
            
            matrix(:,:,1) = parameters%ci(:,:)
            matrix(:,:,2) = parameters%cp(:,:)
            matrix(:,:,3) = parameters%beta(:,:)
            matrix(:,:,4) = parameters%cft(:,:)
            matrix(:,:,5) = parameters%cst(:,:)
            matrix(:,:,6) = parameters%alpha(:,:)
            matrix(:,:,7) = parameters%exc(:,:)
            matrix(:,:,8) = parameters%lr(:,:)
        
        end subroutine parameters_to_matrix
        
        
!%      TODO comment  
        subroutine matrix_to_parameters(matrix, parameters)
            
            implicit none
            
            type(ParametersDT), intent(inout) :: parameters
            real(sp), dimension(size(parameters%cp, 1), &
            & size(parameters%cp, 2), np), intent(in) :: matrix
            
            parameters%ci(:,:) = matrix(:,:,1)
            parameters%cp(:,:) = matrix(:,:,2)
            parameters%beta(:,:) = matrix(:,:,3)
            parameters%cft(:,:) = matrix(:,:,4)
            parameters%cst(:,:) = matrix(:,:,5)
            parameters%alpha(:,:) = matrix(:,:,6)
            parameters%exc(:,:) = matrix(:,:,7)
            parameters%lr(:,:) = matrix(:,:,8)
        
        end subroutine matrix_to_parameters
        
        
!%      TODO comment  
        subroutine vector_to_parameters(vector, parameters)
        
            implicit none
            
            type(ParametersDT), intent(inout) :: parameters
            real(sp), dimension(np), intent(in) :: vector
            
            parameters%ci = vector(1)
            parameters%cp = vector(2)
            parameters%beta = vector(3)
            parameters%cft = vector(4)
            parameters%cst = vector(5)
            parameters%alpha = vector(6)
            parameters%exc = vector(7)
            parameters%lr = vector(8)
        
        end subroutine vector_to_parameters
        
        
!%      TODO comment  
        subroutine set0_parameters(parameters)
        
            implicit none
            
            type(ParametersDT), intent(inout) :: parameters
            
            real(sp), dimension(np) :: vector0
            
            vector0 = 0._sp
            
            call vector_to_parameters(vector0, parameters)
        
        end subroutine set0_parameters
        
        
!%      TODO comment  
        subroutine set1_parameters(parameters)
        
            implicit none
            
            type(ParametersDT), intent(inout) :: parameters
            
            real(sp), dimension(np) :: vector1
            
            vector1 = 1._sp
            
            call vector_to_parameters(vector1, parameters)
        
        end subroutine set1_parameters
        
        
!%      TODO comment
        subroutine hyper_parameters_to_matrix(hyper_parameters, matrix)
        
            implicit none
            
            type(Hyper_ParametersDT), intent(in) :: hyper_parameters
            real(sp), dimension(size(hyper_parameters%cp, 1), &
            & size(hyper_parameters%cp, 2), np), intent(inout) :: matrix

            matrix(:,:,1) = hyper_parameters%ci(:,:)
            matrix(:,:,2) = hyper_parameters%cp(:,:)
            matrix(:,:,3) = hyper_parameters%beta(:,:)
            matrix(:,:,4) = hyper_parameters%cft(:,:)
            matrix(:,:,5) = hyper_parameters%cst(:,:)
            matrix(:,:,6) = hyper_parameters%alpha(:,:)
            matrix(:,:,7) = hyper_parameters%exc(:,:)
            matrix(:,:,8) = hyper_parameters%lr(:,:)
        
        end subroutine hyper_parameters_to_matrix

        
!%      TODO comment
        subroutine matrix_to_hyper_parameters(matrix, hyper_parameters)
        
            implicit none
            
            type(Hyper_ParametersDT), intent(inout) :: hyper_parameters
            real(sp), dimension(size(hyper_parameters%cp, 1), &
            & size(hyper_parameters%cp, 2), np), intent(in) :: matrix
            
            hyper_parameters%ci(:,:) = matrix(:,:,1)
            hyper_parameters%cp(:,:) = matrix(:,:,2)
            hyper_parameters%beta(:,:) = matrix(:,:,3)
            hyper_parameters%cft(:,:) = matrix(:,:,4)
            hyper_parameters%cst(:,:) = matrix(:,:,5)
            hyper_parameters%alpha(:,:) = matrix(:,:,6)
            hyper_parameters%exc(:,:) = matrix(:,:,7)
            hyper_parameters%lr(:,:) = matrix(:,:,8)
        
        end subroutine matrix_to_hyper_parameters
  

!%      TODO comment
        subroutine set0_hyper_parameters(hyper_parameters)
            
            implicit none
            
            type(Hyper_ParametersDT), intent(inout) :: hyper_parameters
            
            real(sp), dimension(size(hyper_parameters%cp, 1), &
            & size(hyper_parameters%cp, 2), np) :: matrix
            
            matrix = 0._sp
            
            call matrix_to_hyper_parameters(matrix, hyper_parameters)
        
        end subroutine set0_hyper_parameters
        
        
!%      TODO comment
        subroutine set1_hyper_parameters(hyper_parameters)
            
            implicit none
            
            type(Hyper_ParametersDT), intent(inout) :: hyper_parameters
            
            real(sp), dimension(size(hyper_parameters%cp, 1), &
            & size(hyper_parameters%cp, 2), np) :: matrix
            
            matrix = 1._sp
            
            call matrix_to_hyper_parameters(matrix, hyper_parameters)
        
        end subroutine set1_hyper_parameters
        
        
!%      TODO comment
        subroutine hyper_parameters_to_parameters(hyper_parameters, &
        & parameters, setup, input_data)
        
            implicit none
            
            type(Hyper_ParametersDT), intent(in) :: hyper_parameters
            type(ParametersDT), intent(inout) :: parameters
            type(SetupDT), intent(in) :: setup
            type(Input_DataDT), intent(in) :: input_data
            
            real(sp), dimension(size(hyper_parameters%cp, 1), &
            & size(hyper_parameters%cp, 2), np) :: hyper_parameters_matrix
            real(sp), dimension(size(parameters%cp, 1), &
            & size(parameters%cp, 2), np) :: parameters_matrix
            real(sp), dimension(size(parameters%cp, 1), &
            & size(parameters%cp, 2)) :: d, dpb
            integer :: i, j
            real(sp) :: a, b
            
            call hyper_parameters_to_matrix(hyper_parameters, hyper_parameters_matrix)
            call parameters_to_matrix(parameters, parameters_matrix)
            
            !% Add mask later here
            !% 1 in dim2 will be replace with k and apply where on Omega
            do i=1, np
            
                parameters_matrix(:,:,i) = hyper_parameters_matrix(1, 1, i)
                
                do j=1, setup%nd
                
                    d = input_data%descriptor(:,:,j)
            
                    a = hyper_parameters_matrix(2 * j, 1, i)
                    b = hyper_parameters_matrix(2 * j + 1, 1, i)
                    dpb = d ** b
                    
                    parameters_matrix(:,:,i) = parameters_matrix(:,:,i) + a * dpb
                
                end do
                
                where (parameters_matrix(:,:,i) .lt. setup%lb_parameters(i))
                    
                    parameters_matrix(:,:,i) = setup%lb_parameters(i)
                
                end where
            
                where (parameters_matrix(:,:,i) .gt. setup%ub_parameters(i))
                    
                    parameters_matrix(:,:,i) = setup%ub_parameters(i)
                
                end where
            
            end do

            call matrix_to_parameters(parameters_matrix, parameters)
        
        end subroutine hyper_parameters_to_parameters
        
end module mwd_parameters

File: smash/solver/_mwd_parameters.py
"""
Module mwd_parameters


Defined at smash/solver/module/mwd_parameters.f90 lines 28-129

This module `mwd_parameters` encapsulates all SMASH parameters.
This module is wrapped and differentiated.

ParametersDT type:

</> Public
======================== =======================================
`Variables`              Description
======================== =======================================
``ci`` Interception parameter [mm] (default: 1) ]0, +Inf[
``cp`` Production parameter [mm] (default: 200) ]0, +Inf[
``beta`` Percolation parameter [-] (default: 1000) ]0, +Inf[
``cft`` Fast transfer parameter [mm] (default: 500) ]0, +Inf[
``cst`` Slow transfer parameter [mm] (default: 500) ]0, +Inf[
``alpha`` Transfer partitioning parameter [-] (default: 0.9) ]0, 1[
``exc`` Exchange parameter [mm/dt] (default: 0) ]-Inf, +Inf[
``lr`` Linear routing parameter [min] (default: 5) ]0, +Inf[
======================== =======================================

contains

[1] ParametersDT_initialise
[2] parameters_to_matrix
[3] matrix_to_parameters
[4] vector_to_parameters
[5] set0_parameters
[6] set1_parameters
"""
from __future__ import print_function, absolute_import, division
from smash.solver._mw_routine import copy_parameters
from smash.solver import _solver
import f90wrap.runtime
import logging

_arrays = {}
_objs = {}

@f90wrap.runtime.register_class("solver.ParametersDT")
class ParametersDT(f90wrap.runtime.FortranDerivedType):
    """
    Type(name=parametersdt)
    
    
    Defined at smash/solver/module/mwd_parameters.f90 lines 32-40
    
    """
    def __init__(self, mesh, handle=None):
        """
        self = Parametersdt(mesh)
        
        
        Defined at smash/solver/module/mwd_parameters.f90 lines 43-69
        
        Parameters
        ----------
        mesh : Meshdt
        
        Returns
        -------
        parameters : Parametersdt
        
        Notes
        -----
        
        ParametersDT initialisation subroutine
        only: sp, np
        only: MeshDT
        """
        f90wrap.runtime.FortranDerivedType.__init__(self)
        result = _solver.f90wrap_parametersdt_initialise(mesh=mesh._handle)
        self._handle = result[0] if isinstance(result, tuple) else result
    
    def __del__(self):
        """
        Destructor for class Parametersdt
        
        
        Defined at smash/solver/module/mwd_parameters.f90 lines 32-40
        
        Parameters
        ----------
        this : Parametersdt
            Object to be destructed
        
        
        Automatically generated destructor for parametersdt
        """
        if self._alloc:
            try:
                _solver.f90wrap_parametersdt_finalise(this=self._handle)
            except:
                pass
    
    @property
    def ci(self):
        """
        Element ci ftype=real(sp) pytype=float
        
        
        Defined at smash/solver/module/mwd_parameters.f90 line 33
        
        """
        array_ndim, array_type, array_shape, array_handle = \
            _solver.f90wrap_parametersdt__array__ci(self._handle)
        if array_handle in self._arrays:
            ci = self._arrays[array_handle]
        else:
            ci = f90wrap.runtime.get_array(f90wrap.runtime.sizeof_fortran_t,
                                    self._handle,
                                    _solver.f90wrap_parametersdt__array__ci)
            self._arrays[array_handle] = ci
        return ci
    
    @ci.setter
    def ci(self, ci):
        self.ci[...] = ci
    
    @property
    def cp(self):
        """
        Element cp ftype=real(sp) pytype=float
        
        
        Defined at smash/solver/module/mwd_parameters.f90 line 34
        
        """
        array_ndim, array_type, array_shape, array_handle = \
            _solver.f90wrap_parametersdt__array__cp(self._handle)
        if array_handle in self._arrays:
            cp = self._arrays[array_handle]
        else:
            cp = f90wrap.runtime.get_array(f90wrap.runtime.sizeof_fortran_t,
                                    self._handle,
                                    _solver.f90wrap_parametersdt__array__cp)
            self._arrays[array_handle] = cp
        return cp
    
    @cp.setter
    def cp(self, cp):
        self.cp[...] = cp
    
    @property
    def beta(self):
        """
        Element beta ftype=real(sp) pytype=float
        
        
        Defined at smash/solver/module/mwd_parameters.f90 line 35
        
        """
        array_ndim, array_type, array_shape, array_handle = \
            _solver.f90wrap_parametersdt__array__beta(self._handle)
        if array_handle in self._arrays:
            beta = self._arrays[array_handle]
        else:
            beta = f90wrap.runtime.get_array(f90wrap.runtime.sizeof_fortran_t,
                                    self._handle,
                                    _solver.f90wrap_parametersdt__array__beta)
            self._arrays[array_handle] = beta
        return beta
    
    @beta.setter
    def beta(self, beta):
        self.beta[...] = beta
    
    @property
    def cft(self):
        """
        Element cft ftype=real(sp) pytype=float
        
        
        Defined at smash/solver/module/mwd_parameters.f90 line 36
        
        """
        array_ndim, array_type, array_shape, array_handle = \
            _solver.f90wrap_parametersdt__array__cft(self._handle)
        if array_handle in self._arrays:
            cft = self._arrays[array_handle]
        else:
            cft = f90wrap.runtime.get_array(f90wrap.runtime.sizeof_fortran_t,
                                    self._handle,
                                    _solver.f90wrap_parametersdt__array__cft)
            self._arrays[array_handle] = cft
        return cft
    
    @cft.setter
    def cft(self, cft):
        self.cft[...] = cft
    
    @property
    def cst(self):
        """
        Element cst ftype=real(sp) pytype=float
        
        
        Defined at smash/solver/module/mwd_parameters.f90 line 37
        
        """
        array_ndim, array_type, array_shape, array_handle = \
            _solver.f90wrap_parametersdt__array__cst(self._handle)
        if array_handle in self._arrays:
            cst = self._arrays[array_handle]
        else:
            cst = f90wrap.runtime.get_array(f90wrap.runtime.sizeof_fortran_t,
                                    self._handle,
                                    _solver.f90wrap_parametersdt__array__cst)
            self._arrays[array_handle] = cst
        return cst
    
    @cst.setter
    def cst(self, cst):
        self.cst[...] = cst
    
    @property
    def alpha(self):
        """
        Element alpha ftype=real(sp) pytype=float
        
        
        Defined at smash/solver/module/mwd_parameters.f90 line 38
        
        """
        array_ndim, array_type, array_shape, array_handle = \
            _solver.f90wrap_parametersdt__array__alpha(self._handle)
        if array_handle in self._arrays:
            alpha = self._arrays[array_handle]
        else:
            alpha = f90wrap.runtime.get_array(f90wrap.runtime.sizeof_fortran_t,
                                    self._handle,
                                    _solver.f90wrap_parametersdt__array__alpha)
            self._arrays[array_handle] = alpha
        return alpha
    
    @alpha.setter
    def alpha(self, alpha):
        self.alpha[...] = alpha
    
    @property
    def exc(self):
        """
        Element exc ftype=real(sp) pytype=float
        
        
        Defined at smash/solver/module/mwd_parameters.f90 line 39
        
        """
        array_ndim, array_type, array_shape, array_handle = \
            _solver.f90wrap_parametersdt__array__exc(self._handle)
        if array_handle in self._arrays:
            exc = self._arrays[array_handle]
        else:
            exc = f90wrap.runtime.get_array(f90wrap.runtime.sizeof_fortran_t,
                                    self._handle,
                                    _solver.f90wrap_parametersdt__array__exc)
            self._arrays[array_handle] = exc
        return exc
    
    @exc.setter
    def exc(self, exc):
        self.exc[...] = exc
    
    @property
    def lr(self):
        """
        Element lr ftype=real(sp) pytype=float
        
        
        Defined at smash/solver/module/mwd_parameters.f90 line 40
        
        """
        array_ndim, array_type, array_shape, array_handle = \
            _solver.f90wrap_parametersdt__array__lr(self._handle)
        if array_handle in self._arrays:
            lr = self._arrays[array_handle]
        else:
            lr = f90wrap.runtime.get_array(f90wrap.runtime.sizeof_fortran_t,
                                    self._handle,
                                    _solver.f90wrap_parametersdt__array__lr)
            self._arrays[array_handle] = lr
        return lr
    
    @lr.setter
    def lr(self, lr):
        self.lr[...] = lr
    
    def __str__(self):
        ret = ['<parametersdt>{\n']
        ret.append('    ci : ')
        ret.append(repr(self.ci))
        ret.append(',\n    cp : ')
        ret.append(repr(self.cp))
        ret.append(',\n    beta : ')
        ret.append(repr(self.beta))
        ret.append(',\n    cft : ')
        ret.append(repr(self.cft))
        ret.append(',\n    cst : ')
        ret.append(repr(self.cst))
        ret.append(',\n    alpha : ')
        ret.append(repr(self.alpha))
        ret.append(',\n    exc : ')
        ret.append(repr(self.exc))
        ret.append(',\n    lr : ')
        ret.append(repr(self.lr))
        ret.append('}')
        return ''.join(ret)

    def copy(self):
        return copy_parameters(self)
    
    _dt_array_initialisers = []
    

def parameters_to_matrix(self, matrix):
    """
    parameters_to_matrix(self, matrix)
    
    
    Defined at smash/solver/module/mwd_parameters.f90 lines 72-84
    
    Parameters
    ----------
    parameters : Parametersdt
    matrix : float array
    
    TODO comment
    """
    _solver.f90wrap_parameters_to_matrix(parameters=self._handle, matrix=matrix)

def matrix_to_parameters(matrix, parameters):
    """
    matrix_to_parameters(matrix, parameters)
    
    
    Defined at smash/solver/module/mwd_parameters.f90 lines 87-99
    
    Parameters
    ----------
    matrix : float array
    parameters : Parametersdt
    
    TODO comment
    """
    _solver.f90wrap_matrix_to_parameters(matrix=matrix, \
        parameters=parameters._handle)

def vector_to_parameters(vector, parameters):
    """
    vector_to_parameters(vector, parameters)
    
    
    Defined at smash/solver/module/mwd_parameters.f90 lines 102-113
    
    Parameters
    ----------
    vector : float array
    parameters : Parametersdt
    
    TODO comment
    """
    _solver.f90wrap_vector_to_parameters(vector=vector, \
        parameters=parameters._handle)

def set0_parameters(self):
    """
    set0_parameters(self)
    
    
    Defined at smash/solver/module/mwd_parameters.f90 lines 116-121
    
    Parameters
    ----------
    parameters : Parametersdt
    
    TODO comment
    """
    _solver.f90wrap_set0_parameters(parameters=self._handle)

def set1_parameters(self):
    """
    set1_parameters(self)
    
    
    Defined at smash/solver/module/mwd_parameters.f90 lines 124-129
    
    Parameters
    ----------
    parameters : Parametersdt
    
    TODO comment
    """
    _solver.f90wrap_set1_parameters(parameters=self._handle)


_array_initialisers = []
_dt_array_initialisers = []

try:
    for func in _array_initialisers:
        func()
except ValueError:
    logging.debug('unallocated array(s) detected on import of module \
        "mwd_parameters".')

for func in _dt_array_initialisers:
    func()
